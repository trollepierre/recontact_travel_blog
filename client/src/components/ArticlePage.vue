<template>
  <div class="page">
    <main class="page__body">
      <div class="page__container">
        <section class="article-page">
          <h1 class="article-page__title">{{ title }}</h1>
          <ul class="chapter__list">
            <li v-for="chapter in chapters" class="chapter__item">
              <chapter-card :chapter="chapter"></chapter-card>
            </li>
          </ul>
        </section>
        <aside>
          <h2 class="article-page__photo-title">{{ $t("hereTheGallery") }}</h2>
          <ul class="photo__list">
            <li v-for="photo in photos" class="photo__item">
              <photo-card :photo="photo"></photo-card>
            </li>
          </ul>
        </aside>
      </div>
    </main>
  </div>
</template>

<script>
  import ChapterCard from '@/components/ChapterCard';
  import PhotoCard from '@/components/PhotoCard';
  import chaptersApi from '@/api/chapters';
  import photosApi from '@/api/photos';
  import translationsService from '@/services/translations';

  export default {
    name: 'ArticlePage',
    components: {
      'chapter-card': ChapterCard,
      'photo-card': PhotoCard,
    },
    data() {
      return {
        chapters: [],
        photos: [],
        title: '',
      };
    },
    mounted() {
      this.getChapters();
      this.getPhotos();
    },
    methods: {
      getChapters() {
        chaptersApi.fetch(this.$route.params.id)
          .then((article) => {
            (this.chapters = article.chapters);
            (this.title = translationsService.getTitle(article));
          });
      },
      getPhotos() {
        photosApi.fetch(this.$route.params.id)
          .then((photos) => {
            (this.photos = photos);
          });
      },
    },
    i18n: {
      messages: {
        fr: {
          hereTheGallery: 'Voici la galerie photo de cet article !',
        },
        en: {
          hereTheGallery: 'Here is the photo gallery of this article',
        },
      },
    },
  };
</script>

<style scoped>
  .page__body {
    display: flex;
    width: 100%;
    padding: 20px 0;
    margin-top: 60px;
    justify-content: center;
  }

  .article-page {
    margin-bottom: 60px;
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }

  .article-page__title {
    font-weight: 300;
    font-size: 24px;
    margin: 0 0 15px;
    max-width: 80%;
  }

  .chapter__list, .photo__list  {
    padding: 0;
    display: flex;
    flex-direction: column;
  }

  .chapter__item, .photo__item  {
    list-style-type: none;
    padding: 0;
    width: 100%;
    margin: 5px;
  }

  @media only screen and (min-width: 640px) {
    .chapter__list, .photo__list  {
      flex-direction: row;
      flex-wrap: wrap;
    }
  }

  @media only screen and (min-width: 640px) {
    .page__container {
      max-width: 544px;
    }
  }

  @media only screen and (min-width: 992px) {
    .page__container {
      max-width: 816px;
    }
  }

  @media only screen and (min-width: 1200px) {
    .page__container {
      max-width: 1088px;
    }
  }

</style>
