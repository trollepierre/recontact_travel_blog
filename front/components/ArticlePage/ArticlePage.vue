<template>
  <main>
    <app-header/>
    <div class="page">
      <section class="article-page">
        <h1 class="article-page__title">
          {{ title || $t("title") }}
        </h1>
        <ul class="chapter__list">
          <li
            v-for="chapter in chapters"
            :key="chapter.imgLink"
            class="chapter__item">
            <chapter-card :chapter="chapter"/>
          </li>
        </ul>
      </section>
      <section class="article-page__forum forum">
        <comments/>
      </section>
      <section
        v-if="!isEmpty(photos)"
        class="article-page__photo-gallery photo-gallery">
        <h2 class="photo-gallery__title">
          {{ $t("hereTheGallery") }}
        </h2>
        <ul class="photo-gallery__list">
          <li
            v-for="photo in photos"
            :key="photo.imgLink"
            class="photo__item">
            <photo-card :photo="photo"/>
          </li>
        </ul>
      </section>
      <footer class="article-page__footer footer-article">
        <button
          class="footer-article__home"
          @click="goToHomePage">
          {{ $t("goToHomePage") }}
        </button>
        <button
          class="footer-article__previous"
          @click="viewPreviousArticle">
          {{ $t("goToPreviousArticle") }}
        </button>
        <button
          class="footer-article__next"
          @click="viewNextArticle">
          {{ $t("goToNextArticle") }}
        </button>
      </footer>
    </div>
  </main>
</template>

<script>
  import { isEmptyPlus as isEmpty } from '../../services'
  import ChapterCard from '../ChapterCard/ChapterCard.vue'
  import PhotoCard from '../PhotoCard/PhotoCard.vue'
  import Comments from '../comments/Comments.vue'
  import chaptersApi from '../../services/api/chapters'
  import photosApi from '../../services/api/photos'
  import translationsService from '../../services/services/translations'
  import logger from '../../services/services/logger-service'
  import AppHeader from '../AppHeader/AppHeader.vue'

  export default {
    name: 'ArticlePage',
    components: {
      ChapterCard,
      Comments,
      PhotoCard,
      AppHeader,
    },
    data() {
      return {
        chapters: [{
          position: 1,
          frTitle: 'Article en cours de chargement',
          enTitle: 'Loading article',
          imgLink: false,
          frText: ['Veuillez patienter quelques secondes'],
          enText: ['Please wait just a second'],
        }],
        photos: [],
        title: '',
        dropboxId: parseInt(this.$route.params.id, 10),
      }
    },
    watch: {
      $route(to) {
        window.scrollTo(0, 0)
        this.dropboxId = parseInt(to.params.id, 10)
        this.fetchArticle()
      },
    },
    mounted() {
      this.fetchArticle()
    },
    methods: {
      fetchArticle() {
        this.getChapters()
        this.getPhotos()
      },
      async getChapters() {
        await chaptersApi.fetch(this.dropboxId)
          .then(article => {
            this.chapters = article.chapters
            const language = this.$store.state.locale
            this.title = translationsService.getTitle(article, language)
          })
          .catch(error => {
            logger.error(error.message)
          })
      },
      getPhotos() {
        photosApi.fetch(this.dropboxId)
          .then(photos => {
            this.photos = photos
          })
      },
      viewPreviousArticle() {
        if (this.dropboxId !== 1) { this.goToArticle(this.dropboxId - 1) }
      },
      viewNextArticle() {
        this.goToArticle(this.dropboxId - 1 + 2)
      },
      goToHomePage() {
        this.$router.push('/')
      },
      goToArticle(idArticle) {
        this.$router.push(`/articles/${idArticle}`)
      },
      isEmpty,
    },
    i18n: {
      silentTranslationWarn: true,
      messages: {
        fr: {
          hereTheGallery: 'Voici la galerie photo de cet article  !',
          goToPreviousArticle: 'Voir l’article précédent',
          goToNextArticle: 'Voir l’article suivant',
          goToHomePage: 'Retour à la page d’accueil',
          title: 'Titre de l’article',
        },
        en: {
          hereTheGallery: 'Here is the photo gallery of this article',
          goToPreviousArticle: 'Read the previous article',
          goToNextArticle: 'Read the next article',
          goToHomePage: 'Go to Home Page',
          title: 'Title of the article',
        },
      },
    },
  }
</script>

<style scoped>
  .page {
    width: 100%;
    padding-top: 70px;
    margin: auto;
  }

  .article-page {
    margin-bottom: 60px;
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }

  .article-page__title {
    font-weight: 300;
    font-size: 32px;
    margin: 0 0 15px;
    max-width: 80%;
  }

  .chapter__list, .photo-gallery__list {
    width: 100%;
    padding: 0;
    display: flex;
    flex-direction: column;
  }

  .chapter__item, .photo__item {
    list-style-type: none;
    padding: 0;
    width: 100%;
    margin: 5px 0;
  }

  .footer-article {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #E6E6E6;
  }

  .footer-article button {
    text-transform: uppercase;
    color: #D14800;
    background: #FFFFFF;
    border: 1px solid #D14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
    max-width: 300px;
  }

  .footer-article button:hover {
    background: #D14800;
    color: #FFFFFF;
  }

  .footer-article button:disabled,
  .footer-article button:active {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (min-width: 640px) {
    .chapter__list, .photo-gallery__list {
      flex-direction: row;
      flex-wrap: wrap;
    }

    .chapter__item, .photo__item {
      margin: 5px;
    }
  }

  @media only screen and (min-width: 640px) {
    .page {
      max-width: 544px;
    }
  }

  @media only screen and (min-width: 992px) {
    .page {
      max-width: 816px;
    }
  }

  @media only screen and (min-width: 1200px) {
    .page {
      max-width: 1088px;
    }
  }

</style>
