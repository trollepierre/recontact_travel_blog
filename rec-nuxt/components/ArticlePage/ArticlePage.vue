<template>
  <div class="page">
    <main class="page__body">
      <div class="page__container">
        <section class="article-page">
          <h1 class="article-page__title">
            {{ title }}
          </h1>
          <ul class="chapter__list">
            <li
              v-for="chapter in chapters"
              :key="chapter.imgLink"
              class="chapter__item">
              <chapter-card :chapter="chapter"/>
            </li>
          </ul>
        </section>
        <section class="article-page__forum forum">
          <comments/>
        </section>
        <section
          v-if="!isEmpty(photos)"
          class="article-page__photo-gallery photo-gallery">
          <h2 class="photo-gallery__title">
            {{ $t("hereTheGallery") }}
          </h2>
          <ul class="photo-gallery__list">
            <li
              v-for="photo in photos"
              :key="photo.imgLink"
              class="photo__item">
              <photo-card :photo="photo"/>
            </li>
          </ul>
        </section>
        <footer class="article-page__footer footer-article">
          <button
            class="footer-article__home"
            @click="goToHomePage">
            {{ $t("goToHomePage") }}
          </button>
          <button
            class="footer-article__previous"
            @click="viewPreviousArticle">
            {{ $t("goToPreviousArticle") }}
          </button>
          <button
            class="footer-article__next"
            @click="viewNextArticle">
            {{ $t("goToNextArticle") }}
          </button>
        </footer>
      </div>
    </main>
  </div>
</template>

<script>
  import { isEmpty } from 'ramda'
  import ChapterCard from '../ChapterCard/ChapterCard.vue'
  import PhotoCard from '../PhotoCard/PhotoCard.vue'
  import Comments from './../comments/Comments.vue'
  import chaptersApi from '../../services/api/chapters'
  import photosApi from '../../services/api/photos'
  import translationsService from '../../services/services/translations'

  export default {
    name: 'ArticlePage',
    components: {
      ChapterCard,
      Comments,
      PhotoCard,
    },
    data() {
      return {
        chapters: [],
        photos: [],
        title: '',
        dropboxId: parseInt(this.$route.params.id, 10),
      }
    },
    watch: {
      $route(to) {
        window.scrollTo(0, 0)
        this.dropboxId = parseInt(to.params.id, 10)
        this.fetchArticle()
      },
    },
    mounted() {
      this.fetchArticle()
    },
    methods: {
      fetchArticle() {
        this.getChapters()
        this.getPhotos()
      },
      async getChapters() {

        this.trackEvent();
        await chaptersApi.fetch(this.dropboxId)
          .then(article => {
            this.chapters = article.chapters;
            const language = this.$store.state.locale
            this.title = translationsService.getTitle(article, language);
          })
      },
      getPhotos() {
        photosApi.fetch(this.dropboxId)
          .then(photos => {
            this.photos = photos
          })
      },
      viewPreviousArticle() {
        this.goToArticle(this.dropboxId - 1)
      },
      viewNextArticle() {
        this.goToArticle(this.dropboxId + 1)
      },
      goToHomePage() {
        this.$router.push('/')
      },
      goToArticle(idArticle) {
        this.$router.push(`/articles/${idArticle}`)
      },
      trackEvent() {
        // this.$ga.event({
        //   eventCategory: 'Article Page',
        //   eventAction: 'read',
        //   eventLabel: `article ${this.$route.params.id} is read`,
        // })
        // TODO fix trackEvent
      },
      isEmpty,
    },
    i18n: {
      messages: {
        fr: {
          hereTheGallery: 'Voici la galerie photo de cet article  !',
          goToPreviousArticle: 'Voir l’article précédent',
          goToNextArticle: 'Voir l’article suivant',
          goToHomePage: 'Retour à la page d’accueil',
        },
        en: {
          hereTheGallery: 'Here is the photo gallery of this article',
          goToPreviousArticle: 'Read the previous article',
          goToNextArticle: 'Read the next article',
          goToHomePage: 'Go to Home Page',
        },
      },
    },
  }
</script>

<style scoped>
  .page__body {
    display: flex;
    width: 100%;
    padding: 20px 0;
    justify-content: center;
  }

  .article-page {
    margin-bottom: 60px;
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }

  .article-page__title {
    font-weight: 300;
    font-size: 24px;
    margin: 0 0 15px;
    max-width: 80%;
  }

  .chapter__list, .photo-gallery__list {
    width: 100%;
    padding: 0;
    display: flex;
    flex-direction: column;
  }

  .chapter__item, .photo__item {
    list-style-type: none;
    padding: 0;
    width: 100%;
    margin: 5px;
  }

  .footer-article {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #e6e6e6;
  }

  .footer-article button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
    max-width: 300px;
  }

  .footer-article button:hover {
    background: #d14800;
    color: #ffffff;
  }

  .footer-article button:disabled,
  .footer-article button:active {
    background: #BDBDBD;
    border-color: #616161;
    color: #FAFAFA;
    cursor: auto;
  }

  @media only screen and (min-width: 640px) {
    .chapter__list, .photo-gallery__list {
      flex-direction: row;
      flex-wrap: wrap;
    }
  }

  @media only screen and (min-width: 640px) {
    .page__container {
      max-width: 544px;
    }
  }

  @media only screen and (min-width: 992px) {
    .page__container {
      max-width: 816px;
    }
  }

  @media only screen and (min-width: 1200px) {
    .page__container {
      max-width: 1088px;
    }
  }

</style>
