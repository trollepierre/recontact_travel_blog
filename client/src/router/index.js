import Vue from 'vue';
import Router from 'vue-router';
import ArticleList from '@/components/ArticleList';
import ArticlePage from '@/components/ArticlePage';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'ArticleList',
      component: ArticleList,
    },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    {
      path: '/admin',
      name: 'ArticleList',
      component: ArticleList,
      props: { adminMode: true },
    },
  ],
});
