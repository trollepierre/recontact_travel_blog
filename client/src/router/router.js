import Vue from 'vue';
import Router from 'vue-router';
import ArticleList from '../components/ArticleList.vue';
import ArticlePage from '../components/ArticlePage.vue';
import SubscriberList from '../components/SubscriberList.vue';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'ArticleList',
      component: ArticleList,
    },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    {
      path: '/admin',
      name: 'ArticleList',
      component: ArticleList,
      props: { adminMode: true },
    },
    {
      path: '/subscriptions',
      name: 'SubscriberList',
      component: SubscriberList,
    },
    { path: '/sub', redirect: '/subscriptions' },
    { path: '/a/:id', redirect: '/articles/:id' },
  ],
});
