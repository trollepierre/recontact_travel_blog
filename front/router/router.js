import Router from 'vue-router'
import Vue from 'vue'
import ArticlePage from '../components/ArticlePage/ArticlePage.vue'

Vue.use(Router)

export const conf = {
  mode: 'history',
  routes: [
    // {
    //   path: '/',
    //   name: 'ArticleList',
    //   component: ArticleList,
    // },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    // {
    //   path: '/admin',
    //   name: 'ArticleList',
    //   component: ArticleList,
    //   props: { adminMode: true },
    // },
    // {
    //   path: '/subscriptions',
    //   name: 'SubscriberList',
    //   component: SubscriberList,
    // },
    { path: '/sub', redirect: '/subscriptions' },
    { path: '/a/:id', redirect: '/articles/:id' },
  ],
}

export default new Router(conf)

