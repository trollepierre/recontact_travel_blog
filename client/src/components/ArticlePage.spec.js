import Vue from 'vue';
import router from '../router/router';
import photosApi from '../api/photos';
import chaptersApi from '../api/chapters';
import ArticlePage from './ArticlePage';
import translationsService from '../services/translations';

xdescribe('Component | ArticlePage.vue', () => {
  let wrapper
  let chapters;
  let photos;
  const title = 'Pierre au pays des';
  const idArticle = 8;

  beforeEach(() => {
    sinon.stub(translationsService, 'getChapterTitle').returns('My title');
    sinon.stub(translationsService, 'getChapterText').returns(['one text']);
    chapters = [
      {
        title: '60 : Pierre avec les webf',
        imgLink: '../assets/toto.jpg',
        text: ['some text'],
      }, {
        title: '61 : Pierre au Koezio',
        imgLink: '/assets/tata.jpg',
        text: ['some text'],
      }, {
        title: '62 : Pierre au Koezio',
        imgLink: '/assets/tata.jpg',
        text: ['some text'],
      },
    ];
    photos = [
      { imgLink: 'url/photo1' },
      { imgLink: 'url/photo2' },
    ];
    sinon.stub(photosApi, 'fetch').resolves(photos);
    sinon.stub(chaptersApi, 'fetch').resolves({ title, chapters });
    const Constructor = Vue.extend(ArticlePage);
    let localVue; localVue = createLocalVue(); wrapper = shallowMount(AppHeader, { localVue, router })
    component.$route.params.id = idArticle;
  });

  afterEach(() => {
    chaptersApi.fetch.restore();
    photosApi.fetch.restore();
    translationsService.getChapterTitle.restore();
    translationsService.getChapterText.restore();
  });

  it('should be named "ArticlePage"', () => {
    expect(wrapper.name()).toEqual('ArticlePage');
  });

  describe('template', () => {
    it('should match snapshot', () => {
      let localVue; localVue = createLocalVue(); wrapper = shallowMount(AppHeader, { localVue })

      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('mounted', () => {
    it('should call chapters api to fetch chapters', () => {
      expect(chaptersApi.fetch).toHaveBeenCalledWith(idArticle);
    });

    it('should call photos api to fetch photos', () => {
      expect(photosApi.fetch).toHaveBeenCalledWith(idArticle);
    });

    it('should save chapters from api in data chapters', () => Vue.nextTick().then(() => {
      expect(wrapper.vm.chapters).toEqual(chapters);
    }));

    it('should save photos from api in data photos', () => Vue.nextTick().then(() => {
      expect(wrapper.vm.photos).toEqual(photos);
    }));
  });

  describe('render', () => {
    it('should render as many chapters as received from the API', () => Vue.nextTick().then(() => {
      const chaptersCards = wrapper.findAll('.chapter-card');
      expect(chaptersCards.length).toEqual(3);
    }));

    it('should have empty chapters in data chapters', () => {
      expect(wrapper.vm.chapters).toEqual([]);
    });

    it('should have empty photos in data photos', () => {
      expect(wrapper.vm.photos).toEqual([]);
    });
  });

  describe('locales', () => {
    const languages = Object.keys(ArticlePage.i18n.messages);

    it('contains 2 languages', () => {
      expect(languages.length).toEqual(2);
      expect(languages).toEqual(['fr', 'en']);
    });

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(ArticlePage.i18n.messages.fr);

        it('contains 4 locales', () => {
          expect(locales.length).toEqual(4);
          expect(locales).toEqual([
            'hereTheGallery',
            'goToPreviousArticle',
            'goToNextArticle',
            'goToHomePage',
          ]);
        });
      });

      describe('en', () => {
        const locales = Object.keys(ArticlePage.i18n.messages.en);

        it('contains 4 locales', () => {
          expect(locales.length).toEqual(4);
          expect(locales).toEqual([
            'hereTheGallery',
            'goToPreviousArticle',
            'goToNextArticle',
            'goToHomePage',
          ]);
        });
      });
    });
  });
});
