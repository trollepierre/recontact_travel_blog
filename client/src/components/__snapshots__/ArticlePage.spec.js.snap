// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | ArticlePage.vue template should match snapshot 1`] = `
<div
  class="page"
>
  <main
    class="page__body"
  >
    <div
      class="page__container"
    >
      <section
        class="article-page"
      >
        <h1
          class="article-page__title"
        >
          
          
        
        </h1>
         
        <ul
          class="chapter__list"
        >
          <li
            class="chapter__item"
          >
            <chapter-card-stub
              chapter="[object Object]"
            />
          </li>
          <li
            class="chapter__item"
          >
            <chapter-card-stub
              chapter="[object Object]"
            />
          </li>
          <li
            class="chapter__item"
          >
            <chapter-card-stub
              chapter="[object Object]"
            />
          </li>
        </ul>
      </section>
       
      <section
        class="article-page__forum forum"
      >
        <comments-stub />
      </section>
       
      <section
        class="article-page__photo-gallery photo-gallery"
      >
        <h2
          class="photo-gallery__title"
        >
          
          hereTheGallery
        
        </h2>
         
        <ul
          class="photo-gallery__list"
        >
          <li
            class="photo__item"
          >
            <photo-card-stub
              photo="[object Object]"
            />
          </li>
          <li
            class="photo__item"
          >
            <photo-card-stub
              photo="[object Object]"
            />
          </li>
        </ul>
      </section>
       
      <footer
        class="article-page__footer footer-article"
      >
        <button
          class="footer-article__home"
        >
          
          goToHomePage
        
        </button>
         
        <button
          class="footer-article__previous"
        >
          
          goToPreviousArticle
        
        </button>
         
        <button
          class="footer-article__next"
        >
          
          goToNextArticle
        
        </button>
      </footer>
    </div>
  </main>
</div>
`;
